lib/Data/Decimate.pm

```diff
@@ -0,0 +1,202 @@
+package Data::Decimate;
+
+use strict;
+use warnings;
+
+use 5.010;
+
+use Exporter qw/import/;
+
+our @EXPORT_OK = qw(decimate);
+
+=head1 NAME
+
+Data::Decimate - A module that allows to decimate a data feed. 
+
+=head1 VERSION
+
+Version 0.01
+
+=head1 SYNOPSIS
+
+  use Data::Decimate qw(decimate);
+
+  my @data = [
+        {epoch  => 1479203101,
+        ...},
+        {epoch  => 1479203102,
+        ...},
+        {epoch  => 1479203103,
+        ...},
+        ...
+        {epoch  => 1479203114,
+        ...},
+        {epoch  => 1479203117,
+        ...},
+        {epoch  => 1479203118,
+        ...},
+        ...
+  ];
+
+  my $output = Data::Decimate::decimate(15, \@data);
+
+  #epoch=1479203114 , decimate_epoch=1479203115
+  print $output->[0]->{epoch};
+  print $output->[0]->{decimate_epoch};
+
+=head1 DESCRIPTION
+
+A module that allows you to resample a data feed
+
+=cut
+
+our $VERSION = '0.01';
+
+=head1 SUBROUTINES/METHODS
+=cut
+
+sub _fill_missing_data {
+    my ($interval, $decimate_data) = @_;
+
+    my @sorted_data = sort { $a <=> $b } keys %$decimate_data;
+    my $first_key   = $sorted_data[0];
+    my $last_key    = $sorted_data[-1];
+
+    if (scalar(@sorted_data)) {
+        for (my $i = $first_key; $i <= $last_key; $i = $i + $interval) {
+            my $data = $decimate_data->{$i};
+
+            if (not $data) {
+                my $data     = $decimate_data->{$i - $interval};
+                my %to_store = %$data;
+                $to_store{decimate_epoch} = $i;
+                $to_store{count}          = 0;
+                $decimate_data->{$i}      = \%to_store;
+            }
+        }
+    }
+
+    return $decimate_data;
+}
+
+=head2 decimate
+
+Decimate a given data based on sampling frequency.
+
+=cut
+
+sub decimate {
+    my ($interval, $data) = @_;
+
+    if (not defined $interval or not defined $data or ref($data) ne "ARRAY") {
+        die "interval and data are required parameters.";
+    }
+
+    my $counter             = 0;
+    my $prev_decimate_epoch = 0;
+    my %decimate_data;
+
+    %decimate_data = map {
+        my $decimate_epoch = ($_->{epoch} % $interval) == 0 ? $_->{epoch} : $_->{epoch} - ($_->{epoch} % $interval) + $interval;
+        $counter = ($decimate_epoch == $prev_decimate_epoch) ? $counter + 1 : 1;
+        $_->{count}          = $counter;
+        $_->{decimate_epoch} = $decimate_epoch;
+        $prev_decimate_epoch = $decimate_epoch;
+        $decimate_epoch => $_
+    } @$data;
+
+    my $tmp = _fill_missing_data($interval, \%decimate_data);
+    my @res = map { $tmp->{$_} } sort { $a <=> $b } keys %$tmp;
+    return \@res;
+}
```

Comment:
I would expect an implementation based on arrays instead of hashes much more efficient. Most probably you can rely on the input being in correct order. Your implementation first thing destroys that order only to re-establish it twice with a relatively expensive `sort`.

This routine (untested) does the same (I think) with only one loop through the input data but it relies on it being in order.

```perl
sub decimate {
    my ($interval, $data) = @_;

    my @res;
    my $el = $data->[0];
    my $de = do {
        use integer;
        (($el->{epoch}+$interval-1) / $interval) * $interval;
    };
    $el->{count} = 1;
    $el->{decimate_epoch} = $de;

    push @res, $el;
    
    for (my $i=1; $i<@$data; $i++) {
        $el = $data->[$i];
        $de = do {
            use integer;
            (($el->{epoch}+$interval-1) / $interval) * $interval;
        };

        # same decimate_epoch
        if ($de == $res[-1]->{decimate_epoch}) {
            $res->[-1]->{count}++;
            next;
        }

        # fill in the gaps if any
        while ($res[-1]->{decimate_epoch} + $interval < $de) {
            my %clone = %{$res[-1]};
            $clone{count} = 0;
            $clone{decimate_epoch} += $interval;
            push @res, \%clone;
        }

         # and finally add the current element
        $el->{count} = 1;
        $el->{decimate_epoch} = $de;
        push @res, $el;
    }

    return \@res;
}
```